File: engine/bridge.py
"""Python client for the Node engine bridge (engine/bridge.js): the real
Showdown Sim.Battle behind the solver's step() interface, via a JSON-lines
child process. Engine-specific glue — model/ never imports this; it gets
step injected (docs/solver_design.md 2.7).

Usage:
    with EngineBridge(FORMAT_ID) as bridge:
        handle, state = bridge.init_battle(full_info_state)
        result = bridge.step(handle, state, my_turn_actions, opp_turn_actions)
        # result.terminal is my-POV +1/-1/0 when the game ended, else None
        # and result.handle/result.state continue the tree

Semantics notes:
- Snapshots are immutable: stepping a handle never mutates it, and each
  step reseeds the engine RNG, so repeated steps from one parent sample
  fresh chance outcomes (what MCCFR's outcome sampling needs).
- Returned FullInfoStates carry REMAINING durations in the field/side
  *_turns fields (engine semantics), unlike BattleState's elapsed-turns
  convention — they only feed my_key/leaf_value/debugging, never re-init.
- Choice strings are built from the PARENT state's team order, which
  matches the engine's side.pokemon order by construction (init orders
  teams [active_left, active_right, bench...] and exports preserve
  engine order thereafter).
- free() releases snapshots Node-side; call it per real decision with the
  handles that solve created, or leak until the process is dropped.
"""

from __future__ import annotations

import json
import subprocess
from pathlib import Path
from typing import NamedTuple, Optional

from schema.battle_state import (
    Action, MoveAction, NoAction, OwnPokemon, Position, SwitchAction, Target, TeamPreviewAction, TurnActions,
)
from schema.full_info_state import FullInfoState, TeamPreviewRootState

_BRIDGE_JS = Path(__file__).resolve().parent / "bridge.js"


class StepResult(NamedTuple):
    handle: Optional[int]      # None when terminal
    state: FullInfoState
    terminal: Optional[float]  # my-POV +1 / -1 / 0, None while ongoing
    errors: list[str]          # engine choice rejections (should stay empty)


def _target_int(target: Target, actor_position: Position) -> int:
    # Same integers as harness/actions.py maps to poke-env: foes 1/2,
    # allies -1/-2, no explicit target 0. Duplicated (6 lines) rather than
    # importing harness/, which is poke-env-specific.
    if target == Target.OPP_LEFT:
        return 1
    if target == Target.OPP_RIGHT:
        return 2
    if target == Target.ALLY:
        return -2 if actor_position == Position.LEFT else -1
    return 0


def _slot_choice(action: Action, team: list[OwnPokemon], actor_position: Position) -> str:
    if isinstance(action, NoAction):
        return "pass"
    if isinstance(action, SwitchAction):
        living_benched = [i for i, m in enumerate(team) if m.position is None and not m.fainted]
        return f"switch {living_benched[action.bench_slot] + 1}"
    if isinstance(action, MoveAction):
        choice = f"move {action.move_slot}"
        target = _target_int(action.target, actor_position)
        if target:
            choice += f" {target}"
        if action.mega:
            choice += " mega"
        return choice
    raise ValueError(f"unknown action type: {action!r}")


def _turn_choice(actions: TurnActions, team: list[OwnPokemon]) -> str:
    return (
        _slot_choice(actions.slot_left, team, Position.LEFT)
        + ", "
        + _slot_choice(actions.slot_right, team, Position.RIGHT)
    )


def _ordered_team(team: list[OwnPokemon]) -> list[OwnPokemon]:
    """[active_left, active_right, bench...] — the order init sends becomes
    the engine's team-preview lead order, so slot 0/1 become active[0]/[1]
    (LEFT/RIGHT). Real Showdown always needs 2 real leads to start a
    battle, so a permanently-empty slot (that side's mon fainted with no
    living replacement - it drops its position, becoming indistinguishable
    from a never-sent bench mon) still needs SOME filler in slot 0/1 to
    seed team preview; init_battle immediately re-faints it via
    applyMonState right after, so which filler is used doesn't matter.

    Without this padding, a side with an empty LEFT and a living RIGHT
    mon would have nothing to put first, so the RIGHT mon would slide into
    slot 0 and get labeled LEFT instead — a silent position flip (caught
    via a real match crash, not by construction).
    """
    left = next((m for m in team if m.position == Position.LEFT), None)
    right = next((m for m in team if m.position == Position.RIGHT), None)
    rest = [m for m in team if m.position is None]
    if left is None:
        left, rest = rest[0], rest[1:]
    if right is None:
        right, rest = rest[0], rest[1:]
    return [left, right] + rest


def _team_preview_choice(action: TeamPreviewAction) -> str:
    # Pure index-ordering math duplicated (not imported) from harness/
    # actions.py::team_preview_action_to_order — engine/ must not depend
    # on harness/, which is poke-env-specific and additionally does
    # protocol framing ("/team " prefix) and a _selected_in_teampreview
    # mutation that don't apply bridge-side. Same "team NNNN" syntax
    # opInit's auto-answer already uses (1-based, no leading slash).
    all_indices = list(range(6))
    bring_rest = [i for i in action.bring if i not in action.lead_order]
    remaining = [i for i in all_indices if i not in action.bring]
    order = action.lead_order + bring_rest + remaining
    return "team " + "".join(str(i + 1) for i in order)


class EngineBridge:
    def __init__(self, format_id: str, node: str = "node"):
        self.format_id = format_id
        self._request_id = 0
        self._proc = subprocess.Popen(
            [node, str(_BRIDGE_JS)],
            stdin=subprocess.PIPE, stdout=subprocess.PIPE,
            text=True, encoding="utf-8", bufsize=1,
        )
        self._rpc({"op": "ping"})  # fail fast if the sim can't load

    def _rpc(self, payload: dict) -> dict:
        self._request_id += 1
        payload["id"] = self._request_id
        self._proc.stdin.write(json.dumps(payload) + "\n")
        self._proc.stdin.flush()
        while True:
            line = self._proc.stdout.readline()
            if not line:
                raise RuntimeError("engine bridge process died")
            try:
                resp = json.loads(line)
            except json.JSONDecodeError:
                continue  # stray non-protocol output
            if resp.get("id") != self._request_id:
                continue
            if not resp.get("ok"):
                raise RuntimeError(f"engine bridge error: {resp.get('error')}")
            return resp

    def init_battle(self, state: FullInfoState) -> tuple[int, FullInfoState]:
        payload_state = {
            "turn": state.turn,
            "field": state.field.model_dump(mode="json"),
            "my_team": [m.model_dump(mode="json") for m in _ordered_team(state.my_team)],
            "opp_team": [m.model_dump(mode="json") for m in _ordered_team(state.opp_team)],
        }
        resp = self._rpc({"op": "init", "format": self.format_id, "state": payload_state})
        return resp["handle"], FullInfoState(**resp["state"])

    def init_team_preview_battle(self, state: TeamPreviewRootState) -> int:
        # Canonical roster order (index i = roster index i) — deliberately
        # NOT run through _ordered_team(): that reordering assumes an
        # already-active-first convention (mid-battle reconstruction),
        # which doesn't apply pre-send-out (nobody's active, all 6 are
        # still just a roster). TeamPreviewAction.bring/lead_order indices
        # are defined against this same canonical order.
        payload_state = {
            "my_team": [m.model_dump(mode="json") for m in state.my_team],
            "opp_team": [m.model_dump(mode="json") for m in state.opp_team],
        }
        resp = self._rpc({"op": "init_team_preview", "format": self.format_id, "state": payload_state})
        return resp["handle"]

    def step_team_preview(self, handle: int, my: TeamPreviewAction, opp: TeamPreviewAction) -> StepResult:
        # Same underlying "step" RPC as step() below, unmodified server-side
        # — Showdown's side.choose() already dispatches "team NNNN" choices
        # generically whenever the restored battle is paused at the
        # teampreview request (see bridge.js's module docstring).
        resp = self._rpc({
            "op": "step",
            "handle": handle,
            "my": _team_preview_choice(my),
            "opp": _team_preview_choice(opp),
        })
        return StepResult(
            handle=resp.get("handle"),
            state=FullInfoState(**resp["state"]),
            terminal=resp.get("terminal"),
            errors=resp.get("errors", []),
        )

    def step(self, handle: int, parent: FullInfoState, my: TurnActions, opp: TurnActions) -> StepResult:
        resp = self._rpc({
            "op": "step",
            "handle": handle,
            "my": _turn_choice(my, parent.my_team),
            "opp": _turn_choice(opp, parent.opp_team),
        })
        return StepResult(
            handle=resp.get("handle"),
            state=FullInfoState(**resp["state"]),
            terminal=resp.get("terminal"),
            errors=resp.get("errors", []),
        )

    def free(self, handles: list[int]) -> None:
        if handles:
            self._rpc({"op": "free", "handles": handles})

    def close(self) -> None:
        if self._proc.poll() is None:
            self._proc.terminate()

    def __enter__(self) -> "EngineBridge":
        return self

    def __exit__(self, *exc) -> None:
        self.close()

File: engine/bridge.js
'use strict';

/**
 * Search-rollout engine bridge: drives Showdown's real Sim.Battle in-process
 * (no server, no websocket) over a JSON-lines stdio protocol, for the
 * solver's step() calls (docs/solver_design.md 2.7).
 *
 * Snapshot model: battles are stored as immutable serialized snapshots
 * (Battle.toJSON strings). Every step restores the parent snapshot into a
 * fresh Battle, RESEEDS ITS RNG (so repeated steps from the same parent
 * sample fresh chance outcomes - MCCFR needs that), advances one full turn,
 * auto-resolves any forced mid-turn replacements with the engine's default
 * choice (= first available switch-in, the documented V1 heuristic), and
 * stores the child as a new snapshot. Parents are never mutated.
 *
 * Root reconstruction (init) builds a real battle from a FullInfoState:
 * teams as set objects (points recovered exactly from final stats via the
 * champions formula's linear inverse: points = stat - base - 20, hp - 75,
 * neutral nature), team preview ordered so actives lead, then direct state
 * mutation (turn, hp, status, boosts, pp, faints, field/side conditions
 * with elapsed->remaining duration conversion, one-mega-per-side flags).
 * Known V1 fidelity gaps, deliberate: volatiles are not reconstructed at
 * the root (Protect streaks, Encore, etc. - they DO evolve correctly for
 * turns simulated below the root), duration items (Light Clay etc.) are
 * assumed absent in remaining-duration math, sleep counters default to 2.
 *
 * Ops: init {format, state} -> {handle, state}
 *      init_team_preview {format, state} -> {handle}  (fresh battle, PAUSED
 *      at the teampreview request - see opInitTeamPreview. Submit the
 *      actual bring/lead choices afterward through the existing step op:
 *      Showdown's side.choose() already dispatches to chooseTeam()
 *      generically whenever requestState === 'teampreview', the same way
 *      it dispatches move/switch choices - opStep needs no changes.)
 *      step {handle, my, opp} -> {handle?, state, terminal?, errors}
 *      free {handles} -> {}
 *      ping -> {}
 * Every request carries an id, echoed in the response. Errors: {ok: false,
 * error}. The exported state mirrors schema/full_info_state.py field names
 * exactly; field/side-condition *_turns carry REMAINING duration (engine
 * semantics), not elapsed - only my_key/leaf/debugging consume them.
 */

const path = require('path');
const readline = require('readline');

const SIM = path.join(__dirname, '..', 'vendor', 'pokemon-showdown', 'dist', 'sim');
const { Battle, Dex } = require(SIM);
const { PRNG } = require(path.join(SIM, 'prng'));

const snapshots = new Map();
let nextHandle = 1;

const WEATHER_TO_ENGINE = { sun: 'sunnyday', rain: 'raindance', sand: 'sandstorm', snow: 'snowscape' };
const ENGINE_TO_WEATHER = {
  sunnyday: 'sun', desolateland: 'sun', raindance: 'rain', primordialsea: 'rain',
  sandstorm: 'sand', hail: 'snow', snowscape: 'snow',
};
const TERRAIN_TO_ENGINE = {
  electric: 'electricterrain', grassy: 'grassyterrain', misty: 'mistyterrain', psychic: 'psychicterrain',
};
const ENGINE_TO_TERRAIN = {
  electricterrain: 'electric', grassyterrain: 'grassy', mistyterrain: 'misty', psychicterrain: 'psychic',
};
const BOOST_FROM_SCHEMA = { atk: 'atk', defense: 'def', spa: 'spa', spd: 'spd', spe: 'spe', acc: 'accuracy', eva: 'evasion' };
const STATS = ['atk', 'def', 'spa', 'spd', 'spe'];
// [schema field, engine condition id, base duration in turns]
const TIMED_SIDE_CONDITIONS = [
  ['reflect_turns', 'reflect', 5], ['light_screen_turns', 'lightscreen', 5],
  ['aurora_veil_turns', 'auroraveil', 5], ['tailwind_turns', 'tailwind', 4],
  ['safeguard_turns', 'safeguard', 5], ['mist_turns', 'mist', 5],
];

function clampDuration(n) {
  return Math.max(1, Math.min(8, n | 0));
}

function buildSet(mon, dex) {
  const species = dex.species.get(mon.species);
  if (!species.exists) throw new Error(`unknown species: ${mon.species}`);
  const base = species.baseStats;
  const evs = { hp: Math.max(0, mon.max_hp - base.hp - 75) };
  for (const s of STATS) evs[s] = Math.max(0, mon.stats[s] - base[s] - 20);
  return {
    name: species.name,
    species: species.name,
    item: mon.item || '',
    ability: mon.ability,
    moves: mon.moves.map(ms => ms.move),
    nature: 'Serious', // neutral: keeps the champions stat formula exactly invertible
    evs,
    ivs: { hp: 31, atk: 31, def: 31, spa: 31, spd: 31, spe: 31 },
    level: mon.level || 50,
  };
}

function applyMonState(p, mon) {
  if (mon.fainted) {
    p.hp = 0;
    p.fainted = true;
    p.status = '';
    p.side.pokemonLeft--;
    return;
  }
  p.hp = Math.max(1, Math.min(mon.hp, p.maxhp));
  if (mon.status && mon.status !== 'none') {
    p.setStatus(mon.status, null, null, true);
    if (p.status !== mon.status) { // immunity from a hypothesized ability - force it anyway
      p.status = mon.status;
      p.statusState = { id: mon.status, target: p };
    }
    if (mon.status === 'slp') p.statusState.time = 2;
    if (mon.status === 'tox') p.statusState.stage = Math.max(1, mon.status_turns || 1);
  } else if (p.status) {
    // True state says no status, but team-preview/send-out setup (e.g. a
    // hypothesized ability or item reacting to something) gave the fresh
    // reconstruction one anyway. Authoritative source always wins - same
    // principle as boosts below, and the same bug class as mega_used: a
    // hidden-info-dependent side effect of reconstruction must never leak
    // through as if it were the true, publicly-observed fact.
    p.status = '';
    p.statusState = { id: '', target: p };
  }
  // Unconditional assignment, not "set if truthy": a genuine 0 must
  // overwrite whatever the engine's own reconstruction produced (e.g.
  // Intimidate firing from an ally's send-out, then a hidden item like
  // Clear Amulet/White Herb - which varies per determinization world -
  // blocking or undoing it). Skipping on falsy 0 let that hidden-item-
  // dependent outcome leak through as if it were the true observed
  // boost, silently diverging the same public fact across worlds.
  for (const [schemaKey, engineKey] of Object.entries(BOOST_FROM_SCHEMA)) {
    p.boosts[engineKey] = (mon.boosts && mon.boosts[schemaKey]) || 0;
  }
  mon.moves.forEach((ms, j) => {
    if (p.moveSlots[j]) {
      p.moveSlots[j].pp = Math.min(ms.pp, p.moveSlots[j].maxpp);
      if (ms.disabled) p.moveSlots[j].disabled = true;
    }
  });
  // Volatiles are otherwise NOT reconstructed at the root (this module's
  // header comment, "Known V1 fidelity gaps") - these two targeted
  // exceptions exist because losing them broke both live play and search
  // rollouts the same way: a mon translated from a real recharge or
  // mid-two-turn-move request (harness/translator.py's
  // _forced_continuation_volatiles, from poke-env's DIRECT must_recharge/
  // preparing_move tracking - model/action_space.py's matching checks key
  // on these exact volatile names) keeps its real moveset with everything
  // disabled, and without the engine's OWN lock volatile it has no idea
  // the mon is locked - it demands a normal choice and rejects whatever
  // gets submitted for an exhausted-bench mon ("Can't pass: ... must make
  // a move"), or wrongly allows an illegal switch otherwise.
  if (mon.volatiles && mon.volatiles.some(v => v.name === 'must_recharge')) {
    // addVolatile (not a raw flag) is what makes Pokemon.getLockedMove()
    // return 'recharge' - the actual mechanism side.ts's chooseMove() uses
    // to force the real move regardless of what we submit, and what makes
    // the engine's own choice validation correctly refuse a switch too.
    p.addVolatile('mustrecharge');
  }
  const twoTurn = mon.volatiles && mon.volatiles.find(v => v.name === 'two_turn_move');
  if (twoTurn) {
    // Same mechanism, generalized to Solar Beam/Fly/Electro Shot-without-
    // Rain/etc.: conditions.ts's twoturnmove.onLockMove returns
    // effectState.move, so addVolatile + directly setting .move (rather
    // than relying on onStart's automatic derivation, which expects to run
    // inside the actual move-execution event context we don't have here)
    // is enough for getLockedMove() to force the real charging move
    // regardless of what slot we submit - same direct-state-mutation
    // reconstruction style as the moveSlots/boosts handling above.
    p.addVolatile('twoturnmove');
    if (p.volatiles.twoturnmove) p.volatiles.twoturnmove.move = twoTurn.data.move;
  }
}

function applySideConditions(side, sc) {
  const source = side.active[0] || side.pokemon[0];
  for (const [schemaKey, id, baseDuration] of TIMED_SIDE_CONDITIONS) {
    if (sc[schemaKey] > 0) {
      side.addSideCondition(id, source);
      if (side.sideConditions[id]) {
        side.sideConditions[id].duration = clampDuration(baseDuration - sc[schemaKey]);
      }
    }
  }
  if (sc.stealth_rock) side.addSideCondition('stealthrock', source);
  if (sc.sticky_web) side.addSideCondition('stickyweb', source);
  for (let i = 0; i < (sc.spikes_layers || 0); i++) side.addSideCondition('spikes', source);
  for (let i = 0; i < (sc.toxic_spikes_layers || 0); i++) side.addSideCondition('toxicspikes', source);
  // Tracked as an explicit flag, not inferred from canMegaEvo: whether any
  // CURRENT mon happens to hold a legal stone varies by determinized world
  // (a hidden bench mon's sampled item differs per world), so canMegaEvo
  // is world-dependent even before any mega has actually happened - using
  // it for mega_used leaked hidden info into the public key and broke the
  // solver's world-independence contract. A plain boolean here survives
  // Battle.toJSON/fromJSON via state.ts's generic own-key walk, so it
  // persists correctly across snapshot/restore through the whole tree.
  side.initialMegaUsed = !!sc.mega_used;
  if (sc.mega_used) {
    for (const p of side.pokemon) p.canMegaEvo = false;
  }
}

function applyField(battle, field) {
  const source = battle.p1.active[0] || battle.p1.pokemon[0];
  if (field.weather && field.weather !== 'none') {
    battle.field.setWeather(WEATHER_TO_ENGINE[field.weather], source);
    battle.field.weatherState.duration = clampDuration(5 - (field.weather_turns || 0));
  } else {
    battle.field.clearWeather(); // squash anything a start-of-battle ability set
  }
  if (field.terrain && field.terrain !== 'none') {
    battle.field.setTerrain(TERRAIN_TO_ENGINE[field.terrain], source);
    battle.field.terrainState.duration = clampDuration(5 - (field.terrain_turns || 0));
  } else {
    battle.field.clearTerrain();
  }
  if (field.trick_room_turns > 0) {
    battle.field.addPseudoWeather('trickroom', source);
    if (battle.field.pseudoWeather['trickroom']) {
      battle.field.pseudoWeather['trickroom'].duration = clampDuration(5 - field.trick_room_turns);
    }
  }
  if (field.gravity_turns > 0) {
    battle.field.addPseudoWeather('gravity', source);
    if (battle.field.pseudoWeather['gravity']) {
      battle.field.pseudoWeather['gravity'].duration = clampDuration(5 - field.gravity_turns);
    }
  }
  applySideConditions(battle.p1, field.my_side);
  applySideConditions(battle.p2, field.opp_side);
}

function exportMoves(p) {
  // Pokemon.getMoveRequestData() is the SAME call the engine's own request
  // generation uses - not a reimplementation. It narrows `moves` to a
  // single synthetic entry ({id:'recharge'} or {id:'struggle'}) or a
  // single real entry (a two-turn move like Solar Beam mid-charge) when
  // the mon has no free choice, instead of the normal 4-move list with
  // per-move `disabled` flags. Real pp/maxpp aren't included on locked/
  // synthetic entries (they're not being spent from a normal moveslot),
  // so those fall back to a placeholder of 1 - just needs to be nonzero
  // so the pp<=0 filter in propose_slot_actions doesn't wrongly drop the
  // only legal action.
  const req = p.getMoveRequestData();
  // p.trapped (NOT req.trapped) for the trapped flag: getMoveRequestData()
  // deliberately HIDES trapped-status from the exported request when the
  // cause is an undiscovered ability (Shadow Tag/Arena Trap) - it sets
  // p.trapped to the string 'hidden' rather than true, and its own export
  // logic uses a strict `=== true` check that excludes it, matching what
  // a real player genuinely wouldn't be told yet. That's correct for the
  // live protocol (poke-env's battle.trapped mirrors it) but wrong here:
  // search reasons inside an already-fully-determinized world, so there's
  // no information left to hide from ourselves. Confirmed via a
  // standalone diagnostic against the real engine (not guessed): the
  // ability/event system itself was never broken, only this one read.
  return {
    moves: req.moves.map(m => {
      const real = p.moveSlots.find(ms => ms.id === m.id);
      return {
        move: m.id,
        pp: real ? real.pp : 1,
        max_pp: real ? real.maxpp : 1,
        disabled: !!m.disabled,
      };
    }),
    trapped: !!p.trapped,
  };
}

function exportMon(p, side) {
  const activeIdx = side.active.indexOf(p);
  const position = activeIdx === 0 ? 'left' : activeIdx === 1 ? 'right' : null;
  const boosts = {
    atk: p.boosts.atk, defense: p.boosts.def, spa: p.boosts.spa, spd: p.boosts.spd,
    spe: p.boosts.spe, acc: p.boosts.accuracy, eva: p.boosts.evasion,
  };
  const stats = {};
  for (const s of STATS) stats[s] = p.storedStats[s];
  const { moves, trapped } = exportMoves(p);
  return {
    species: p.species.id,
    level: p.level,
    position: p.fainted ? null : position,
    fainted: !!p.fainted,
    hp: p.hp,
    max_hp: p.maxhp,
    status: (p.status && p.status !== 'fnt') ? p.status : 'none', // fnt is carried by `fainted`, not a schema Status
    status_turns: (p.statusState && p.statusState.time) || null,
    stats,
    boosts,
    ability: p.ability,
    item: p.item || null,
    moves,
    mega_activated: !!p.species.isMega,
    volatiles: Object.keys(p.volatiles).map(name => ({ name })),
    trapped,
  };
}

function conditionDuration(sc, id) {
  return (sc[id] && sc[id].duration) || 0;
}

function exportSide(side) {
  const sc = side.sideConditions;
  const out = {
    stealth_rock: !!sc.stealthrock,
    spikes_layers: (sc.spikes && sc.spikes.layers) || 0,
    toxic_spikes_layers: (sc.toxicspikes && sc.toxicspikes.layers) || 0,
    sticky_web: !!sc.stickyweb,
    // "already used" is set explicitly at init (side.initialMegaUsed) and
    // otherwise detected by species.isMega, which is permanent once a mega
    // evolution actually happens (even after the mon later faints, per the
    // champions mod's formeChange override) - NOT by scanning canMegaEvo,
    // which is contaminated by per-world item-holding uncertainty (see
    // applySideConditions above).
    mega_used: !!side.initialMegaUsed || side.pokemon.some(p => !!p.species.isMega),
    tera_used: false,
  };
  for (const [schemaKey, id] of TIMED_SIDE_CONDITIONS.map(([k, i]) => [k, i])) {
    out[schemaKey] = conditionDuration(sc, id);
  }
  return out;
}

function exportState(battle) {
  const f = battle.field;
  return {
    turn: battle.turn,
    field: {
      weather: ENGINE_TO_WEATHER[f.weather] || 'none',
      weather_turns: (f.weatherState && f.weatherState.duration) || 0,
      terrain: ENGINE_TO_TERRAIN[f.terrain] || 'none',
      terrain_turns: (f.terrainState && f.terrainState.duration) || 0,
      trick_room_turns: (f.pseudoWeather['trickroom'] && f.pseudoWeather['trickroom'].duration) || 0,
      gravity_turns: (f.pseudoWeather['gravity'] && f.pseudoWeather['gravity'].duration) || 0,
      my_side: exportSide(battle.p1),
      opp_side: exportSide(battle.p2),
    },
    my_team: battle.p1.pokemon.map(p => exportMon(p, battle.p1)),
    opp_team: battle.p2.pokemon.map(p => exportMon(p, battle.p2)),
  };
}

function store(battle) {
  const handle = nextHandle++;
  snapshots.set(handle, JSON.stringify(battle.toJSON()));
  return handle;
}

function resolveForcedSwitches(battle) {
  // A mon fainted mid-turn: the engine wants replacements before the next
  // full turn. Default choice = first available switch-in (the V1 in-search
  // heuristic; the live game solves these as their own decision instead).
  let guard = 0;
  while (!battle.ended && battle.requestState === 'switch' && guard++ < 12) {
    battle.makeChoices();
  }
}

function opInit(req) {
  const format = Dex.formats.get(req.format, true);
  const dex = Dex.forFormat(format);
  const battle = new Battle({
    formatid: format.id,
    strictChoices: false,
    debug: false,
    p1: { name: 'me', team: req.state.my_team.map(m => buildSet(m, dex)) },
    p2: { name: 'opp', team: req.state.opp_team.map(m => buildSet(m, dex)) },
  });
  if (battle.requestState === 'teampreview') {
    const order = n => 'team ' + '123456'.slice(0, n);
    battle.makeChoices(order(req.state.my_team.length), order(req.state.opp_team.length));
  }
  battle.turn = req.state.turn;
  req.state.my_team.forEach((mon, i) => applyMonState(battle.p1.pokemon[i], mon));
  req.state.opp_team.forEach((mon, i) => applyMonState(battle.p2.pokemon[i], mon));
  applyField(battle, req.state.field);
  return { handle: store(battle), state: exportState(battle) };
}

function opInitTeamPreview(req) {
  // A genuinely fresh battle for the team-preview decision itself - NOT a
  // mid-battle reconstruction (that's opInit: auto-answers team preview
  // with everyone in submitted order, then hand-mutates turn/hp/status/
  // boosts to match an already-in-progress position). Here nothing is
  // mutated or recomputed: req.state.my_team/opp_team are the full 6-mon
  // rosters (schema/full_info_state.py's TeamPreviewRootState), buildSet
  // is reused unchanged (already generic over roster size/position), and
  // the battle is returned exactly where `new Battle(...)` naturally
  // leaves it - paused at requestState 'teampreview' - so the caller can
  // branch multiple different bring/lead choices off the same snapshot
  // via the existing step op.
  const format = Dex.formats.get(req.format, true);
  const dex = Dex.forFormat(format);
  const battle = new Battle({
    formatid: format.id,
    strictChoices: false,
    debug: false,
    p1: { name: 'me', team: req.state.my_team.map(m => buildSet(m, dex)) },
    p2: { name: 'opp', team: req.state.opp_team.map(m => buildSet(m, dex)) },
  });
  return { handle: store(battle) };
}

function opStep(req) {
  const serialized = snapshots.get(req.handle);
  if (!serialized) throw new Error(`unknown handle ${req.handle}`);
  const battle = Battle.fromJSON(serialized);
  battle.resetRNG(PRNG.generateSeed());

  const errors = [];
  if (!battle.p1.choose(req.my)) {
    errors.push(`p1 "${req.my}": ${(battle.p1.choice && battle.p1.choice.error) || 'invalid'}`);
  }
  if (!battle.p2.choose(req.opp)) {
    errors.push(`p2 "${req.opp}": ${(battle.p2.choice && battle.p2.choice.error) || 'invalid'}`);
  }
  battle.makeChoices(); // auto-completes any failed side, then commits
  resolveForcedSwitches(battle);

  const out = { state: exportState(battle), errors };
  if (battle.ended) {
    out.terminal = battle.winner === 'me' ? 1 : battle.winner === 'opp' ? -1 : 0;
  } else {
    out.handle = store(battle);
  }
  return out;
}

const rl = readline.createInterface({ input: process.stdin, crlfDelay: Infinity });
rl.on('line', line => {
  if (!line.trim()) return;
  let req;
  try {
    req = JSON.parse(line);
  } catch {
    return;
  }
  let resp;
  try {
    if (req.op === 'init') resp = opInit(req);
    else if (req.op === 'init_team_preview') resp = opInitTeamPreview(req);
    else if (req.op === 'step') resp = opStep(req);
    else if (req.op === 'free') { for (const h of req.handles) snapshots.delete(h); resp = {}; }
    else if (req.op === 'ping') resp = { snapshots: snapshots.size };
    else throw new Error(`unknown op ${req.op}`);
    resp.ok = true;
  } catch (err) {
    resp = { ok: false, error: err.stack || String(err) };
  }
  resp.id = req.id;
  process.stdout.write(JSON.stringify(resp) + '\n');
});
